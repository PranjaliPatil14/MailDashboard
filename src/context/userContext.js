import React, { useEffect, useState } from "react";
import * as PropTypes from "prop-types";
import useLocalStorage, { USER } from "../utilities/useLocalStorage";

const initialUserContextValue = {
  users: [],
  loggedInUser: {
    id: 0,
    name: "",
    role: "",
    profileImage: "",
  },
  setLoggedInUser: () => {},
  setUsers: () => {},
};

export const UserContext = React.createContext(initialUserContextValue);

const UserContextProvider = ({ children }) => {
  const [users, setUsers] = useState([]);
  const [loggedInUser, setLoggedInUser] = useState({});
  const { value } = useLocalStorage(USER);

  useEffect(() => {
    if (!users.length) {
      setUsers(value);
    }
  }, [users.length, value]);

  return (
    <UserContext.Provider
      value={{ users, setUsers, loggedInUser, setLoggedInUser }}
    >
      {children}
    </UserContext.Provider>
  );
};

UserContextProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

export default UserContextProvider;
