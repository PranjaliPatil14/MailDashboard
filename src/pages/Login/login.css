@import "../../assets/styles/colors.css";

.login {
    margin: 10% auto 0 auto;
    width: 500px;
    display: flex;
    flex-direction: column;
    text-align: center;
    vertical-align: middle;
}

.username {
    padding: 20px;
}

.password {
    padding: 20px;
    margin-top: 10px;
}

.login-button {
    padding: 20px;
    background: var(--green);
    border-radius: 5px;
    margin-top: 10px;
    color: white;
    font-size: 16px;
    font-weight: bold;
}

.welcome-text {
    color: var(--green);
}
