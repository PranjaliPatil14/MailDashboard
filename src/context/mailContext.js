import React, { useContext, useEffect, useReducer } from "react";
import * as PropTypes from "prop-types";
import useLocalStorage, { MAIL } from "../utilities/useLocalStorage";
import { UserContext } from "./userContext";
import {
  mailReducer,
  SET_LOGGED_IN_USER_MAIL,
  SET_VALUES,
} from "./mailReducer";

const initialMailContextValue = {
  mails: [],
  refresh: true,
  loggedInUserMails: [],
  dispatch: () => {},
};

export const MailContext = React.createContext(initialMailContextValue);

const MailContextProvider = ({ children }) => {
  const [state, dispatch] = useReducer(mailReducer, initialMailContextValue);
  const { value } = useLocalStorage(MAIL);
  const { loggedInUser } = useContext(UserContext);

  useEffect(() => {
    if (state.refresh) {
      dispatch({ type: SET_VALUES, data: value });
    }
  }, [value, state.refresh]);

  useEffect(() => {
    dispatch({ type: SET_LOGGED_IN_USER_MAIL, data: loggedInUser.id });
  }, [loggedInUser, value, state.mails]);

  useEffect(() => {
    localStorage.setItem(MAIL, JSON.stringify(state.mails));
  }, [state.mails]);

  return (
    <MailContext.Provider value={{ ...state, dispatch }}>
      {children}
    </MailContext.Provider>
  );
};

MailContextProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

export default MailContextProvider;
