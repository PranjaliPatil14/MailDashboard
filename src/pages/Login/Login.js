import React, { useContext, useState } from "react";
import Button from "../../components/button/Button";
import { UserContext } from "../../context/userContext";
import "./login.css";

const Login = () => {
  const [username, setUserName] = useState();
  const [password, setPassword] = useState();
  const { users, setLoggedInUser } = useContext(UserContext);
  const onSubmit = () => {
    const user = users.find(({ email }) => email === username);
    if (user?.password === password) {
      setLoggedInUser(user);
    }
  };
  return (
    <div className="login">
      <h1 className="welcome-text">Welcome</h1>
      <input
        className="username"
        type="text"
        placeholder="Enter your email"
        value={username}
        onChange={({ target }) => {
          setUserName(target.value);
        }}
      />
      <input
        className="password"
        type="password"
        placeholder="Enter your password"
        value={password}
        onChange={({ target }) => {
          setPassword(target.value);
        }}
      />
      <Button onClick={onSubmit} className="login-button">
        Login
      </Button>
    </div>
  );
};

export default Login;
