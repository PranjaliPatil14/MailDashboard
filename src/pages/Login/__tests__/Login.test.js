import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Login from "../Login";
import { UserContext } from "../../../context/userContext";

describe("login", () => {
  const setLoggedInUser = jest.fn();
  const users = [{ email: "abc@gmail.com", password: "abc" }];
  let component;
  beforeEach(() => {
    component = render(
      <UserContext.Provider value={{ setLoggedInUser, users }}>
        <Login />
      </UserContext.Provider>
    );
  });
  test("should call setLoggedInUser if password and username is correct", () => {
    expect(component.asFragment()).toMatchSnapshot();
    fireEvent.change(component.getByTestId("username"), {
      target: {
        value: "abc@gmail.com",
      },
    });
    fireEvent.change(component.getByTestId("password"), {
      target: {
        value: "abc",
      },
    });

    fireEvent.click(component.getByText("Login"));

    expect(setLoggedInUser).toHaveBeenCalled();
  });

  test("should not call setLoggedInUser if password and username is incorrect", () => {
    fireEvent.change(component.getByTestId("username"), {
      target: {
        value: "abc@gmail.com",
      },
    });
    fireEvent.change(component.getByTestId("password"), {
      target: {
        value: "wrong password",
      },
    });

    fireEvent.click(component.getByText("Login"));

    expect(setLoggedInUser).not.toHaveBeenCalled();
  });
});
