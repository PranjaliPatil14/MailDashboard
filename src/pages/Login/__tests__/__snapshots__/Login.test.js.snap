// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login should call setLoggedInUser if password and username is correct 1`] = `
<DocumentFragment>
  <div
    class="login"
  >
    <h1
      class="welcome-text"
    >
      Welcome
    </h1>
    <input
      class="username"
      data-testid="username"
      placeholder="Enter your email"
      type="text"
      value=""
    />
    <input
      class="password"
      data-testid="password"
      placeholder="Enter your password"
      type="password"
      value=""
    />
    <button
      class="button login-button"
      data-testid=""
      type="submit"
    >
      Login
    </button>
  </div>
</DocumentFragment>
`;
